### PR TITLE
Separate implementation and fix types

### PR DESCRIPTION
**Pull request by hannahhoward**
# Goals

Per discussions with @aschmahmann, this PR moves the implementation to a separate package from the interfaces. It also makes the fetching functions take ipld.Node's instead of parsed selectors.

# Implementation

- make an impl folder with a blockservice implementation subfolder -- cause eventually there may be other implementations that work directly with bitswap and graphsync
- move the traversal helpers into the helpers folder
- change the functions to take ipld.Node's as parameters
- also adds circle ci config! wow that was late.

### CODE REVIEW

File: fetcher.go [hunk 1]
```diff
@@ -1,37 +1,19 @@
...
-
 // Fetcher is an interface for reading from a dag. Reads may be local or remote, and may employ data exchange
 // protocols like graphsync and bitswap
 type Fetcher interface {
 	// NodeMatching traverses a node graph starting with the provided node using the given selector and possibly crossing
 	// block boundaries. Each matched node is passed as FetchResult to the callback. Errors returned from callback will
 	// halt the traversal. The sequence of events is: NodeMatching begins, the callback is called zero or more times
 	// with a FetchResult, then NodeMatching returns.
-	NodeMatching(context.Context, ipld.Node, selector.Selector, FetchCallback) error
+	NodeMatching(context.Context, ipld.Node, ipld.Node, FetchCallback) error
```
Comment (willscott) [inline fetcher.go:16]: this isn't just any node though, it's still a selector? is there a way to make that signature clearer while still being efficient @warpfork? 

At least we probably want a comment on it / usage example

**Review by willscott — APPROVED** [review]

File: fetcher.go (hunk 1 above)
Comment (hannahhoward) [reply fetcher.go:16]: technically though it is any ipld.Node that parses to a selector.

Comment (willscott) [reply fetcher.go:16]: i think what I'm noting is that this interface method signature now provides no indication of which argument is the root and which is the selector. that needs to exist as argument names / comment if it isn't clear from the type signature.

Comment (hannahhoward) [reply fetcher.go:16]: yea totally :)

Comment (hannahhoward) [reply fetcher.go]: updated somewhat, hopefully in a way thats clear

**Review by willscott — APPROVED** [review]

**Pull request merged**